### PR TITLE
Real time dashboard updates from the API

### PR DESCRIPTION
**Pull request by fulljames**
# BM-242
- Created a socket simulator which:
  - is used by appending `?simulator` to the dashboard URL, otherwise the app attempts a connection to the back end
  - can respond to the initial socket GET request for centres data, and individual model subscriptions to listen to model updates
  - on subscriptions it emits new data on a randomly set interval
- Created a centre view which can be instantiated _n_ times on the main dashboard
- Tests for calculated properties on centre model and socket events
- Clean-up of obsolete code / JSON / tests / etc

### CODE REVIEW

File: assets/js/socketSimulator.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
+var _ = require('underscore');
+
+var socket = require('socket-io-mock');
+var server = new socket();
+
+server.socketClient.connected = true;
+
+server.on('get', function(payload) {
+  //console.log(payload);
```
Comment (chrisns) [inline assets/js/socketSimulator.js]: can be removed

Comment (chrisns) [inline assets/js/socketSimulator.js]: remove commented code

File: gulpfile.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ gulp.task('templates', function() {
 
 gulp.task('copy', function() {
   gulp.src('./assets/json/*.json').pipe(gulp.dest('./dist'));
-  gulp.src('./node_modules/govuk_template_mustache/assets/images/*/**').pipe(gulp.dest('./dist/images'));
+  gulp.src('./node_modules/govuk_template_mustache/assets/images/*.**').pipe(gulp.dest('./dist/images'));
```
Comment (chrisns) [inline gulpfile.js:65]: i suspect you probably want `**/*` not `*.**`?

File: package.json [hunk 3]
```diff
@@ -25,6 +25,9 @@
     "jquery": "^2.1.4",
     "jsdom": "^6.5.1",
     "mocha": "^2.3.3",
+    "sails.io.js": "^0.11.7",
+    "socket-io-mock": "^1.0.1",
+    "socket.io-client": "^1.3.7",
     "underscore": "^1.8.3",
```
Comment (chrisns) [inline package.json]: any reason for underscore over lodash?
typically lodash is more performant

File: test/models.test.js [hunk 4]
```diff
@@ -50,10 +52,74 @@ describe('collections', function () {
...
+
+      io = sailsIOClient(this.fakeio(this.server));
+      io.sails.autoConnect = false;
+
+      this.socket0 = io.sails.connect();
+    });
     it('should be instantiated', function() {
       var model = new models.Centre();
 
       expect(model).not.to.be(undefined);
```
Comment (chrisns) [inline test/models.test.js]: in general unit tests should have one assert.
you should also return the expect so that mocha picks it all up explicity

File: test/models.test.js [hunk 5]
```diff
@@ -50,10 +52,74 @@ describe('collections', function () {
...
+      expect(model.get('reserved')).to.be(0);
+
+    });
+
+    it('should respond to socket events', function(done) {
+      var model = new models.Centre([], { socket: this.socket0 });
+
+      model.socket.on('message', function(payload) {
+        expect(payload).to.be('test message');
+        done();
```
Comment (chrisns) [inline test/models.test.js]: ideally do this with promises rather than passing done which will allow you to ultimately return the expect back up to mocha easily

File: gulpfile.js (hunk 2 above)
Comment (sassquad) [inline gulpfile.js:65]: Interesting you said that. For some reason, Gulp was pulling images from a different folder when I used `**/*`. I therefore changed it to `*.**` so it could capture just any file in that specific images folder (there are no subfolders there, all but one file are .png, the only other file being an .ico file.

File: test/models.test.js (hunk 4 above)
Comment (sassquad) [inline test/models.test.js]: In Stephen's absence, can you clarify your comment @chrisns ? Are you referring to the last four lines of the above snippet?

Something like the following? Feel free to correct me!

```
it('should be instantiated', function(done,error) {
   var model = new models.Centre();
   expect(model).to.be.ok();
   expect(model).not.to.be(undefined);
...
```

Comment (chrisns) [inline test/models.test.js]: huh? no, there should be just one expect per `it`, and it should be what the `it()` returns which allows the expects to bubble up through the reporting

``` javascript
it('should be instantiated', function() {
  var model = new models.Centre();
  return expect(model).to.be.ok();
});
```

Comment (sassquad) [inline test/models.test.js]: Thanks @chrisns for the clarification, make the tests as verbose as possible, rather than bundle them together as above.

Suggesting the return was my alternative code snippet, should have gone with that instead of what I submitted. Thanks again.

File: assets/js/socketSimulator.js (hunk 1 above)
Comment (sassquad) [inline assets/js/socketSimulator.js]: I've removed this line, and performed a direct commit via Github into the same PR.

File: test/models.test.js (hunk 4 above)
Comment (sassquad) [inline test/models.test.js]: I've refined the tests for this block of code in a new commit.

**Comment by sassquad** [comment]
There's a botched commit in there (6a6f1f4) that contains nothing.

File: assets/js/app.js [hunk 6]
```diff
@@ -1,18 +1,54 @@
...
+      });
+
+      thisView.model.set(item);
+    })
+
+  });
+
+  var $container = $('section.tabstuff');
+
+  if ($container.find('.js-tabs').length) {
```
Comment (chrisns) [inline assets/js/app.js]: you could probably use a better selector rather than having an if, you should also be explicit rather than implicitly relying on 0 to be falsey

File: assets/js/app.js [hunk 7]
```diff
@@ -1,18 +1,54 @@
...
+        el: '#item'+idx,
+        socket: socket0
+      });
+
+      thisView.model.set(item);
+    })
+
+  });
+
+  var $container = $('section.tabstuff');
```
Comment (chrisns) [inline assets/js/app.js]: `tabsuff` is probably not a great selector, is there no better semantic name?

File: assets/js/app.js [hunk 8]
```diff
@@ -1,18 +1,54 @@
...
 
-  new view(options);
+  io = sailsIOClient(socketio);
+
+  io.sails.autoConnect = false;
+  if (!simulatorParam) io.sails.url = 'http://localhost:8080';
+
+  socket0 = io.sails.connect();
+
+  socket0.get('/centre', function serverResponded(payload) {
```
Comment (chrisns) [inline assets/js/app.js:30]: this won't handle centres being added or removed while the socket is up, can this not databind instead rather than init containers for each centre on startup?

File: assets/js/models/centre.js [hunk 9]
```diff
@@ -1,5 +1,48 @@
 var Backbone = require('backbone');
+var _ = require('underscore');
```
Comment (chrisns) [inline assets/js/models/centre.js:2]: you'll generally find that lodash outperforms underscore

File: assets/js/models/centre.js [hunk 10]
```diff
@@ -1,5 +1,48 @@
...
+      return memo + item.available;
+    }, 0));
+  },
+
+  subscribeToCentre: function() {
+    var self = this;
+    this.socket.on('centre_id/' + this.get('centre_id'), function(payload) {
+      self.set(payload);
+    });
+    this.socket._raw.emit('subscribe', this.get('centre_id'));
```
Comment (chrisns) [inline assets/js/models/centre.js]: from what I can see you should maybe be able to do `this.socket.subscribe()`

File: assets/templates/bed.html [hunk 11]
```diff
@@ -1 +1 @@
-<li><span class='bed-count-figure'>{{available}}</span> <span class='bed-count-label'>{{type}} beds available</span><input class='button' type='submit' value='Book 1 bed'><a href='#'>Ring-fence beds</a></li>
+<li><span class="bed-count-figure">{{available}}</span> <span class="bed-count-label">{{type}} beds available</span></li>
```
Comment (chrisns) [inline assets/templates/bed.html]: should probably format through something that'll figure out the plural so, `1 bed` vs `0 beds`, `2 beds`

File: assets/js/models/centre.js (hunk 10 above)
Comment (fulljames) [inline assets/js/models/centre.js]: @chrisns Ah yep, that's what I was looking for. Also realised that we are only emitting an event here to tell the simulator to emit events, when in reality there would just be an existing stream to listen to, so I will revise this a bit.

File: assets/js/app.js (hunk 6 above)
Comment (sassquad) [inline assets/js/app.js]: Thanks @chrisns for the comments - given the revised MVP, the tabs are effectively redundant, I'll therefore be removing the code. This is admittedly scrappy code, so I agree with you.

File: assets/js/app.js (hunk 7 above)
Comment (sassquad) [inline assets/js/app.js]: @chrisns - of course there is. We're Agile, right?

File: assets/templates/bed.html (hunk 11 above)
Comment (sassquad) [inline assets/templates/bed.html]: Indeed, will look at this as well, @chrisns

File: assets/js/app.js (hunk 8 above)
Comment (fulljames) [inline assets/js/app.js:30]: @chrisns I'll think about this separately. It's maybe a bit of an edge case (for now) and a little outside the scope of "make real time happen" for this user story.

File: test/models.test.js [hunk 12]
```diff
@@ -1,59 +1,227 @@
...
+
+        expect(female_beds.available).to.be(55);
+        return done();
+      });
+
+      model.set('beds',data);
+    });
+
+    it('should check whether there is only 1 female bed', function(done) {
+      var model = new models.Centre();
```
Comment (chrisns) [inline test/models.test.js:148]: you could move all the `model` and `data` definitions into a `beforeEach()` which would reduce the repetition between your tests here.
